src/librustc_data_structures/obligation_forest/mod.rs

```diff
@@ -42,34 +66,40 @@ pub struct ObligationForest<O, T> {
     /// at a higher index than its parent. This is needed by the
     /// backtrace iterator (which uses `split_at`).
     nodes: Vec<Node<O>>,
-    trees: Vec<Tree<T>>,
-    snapshots: Vec<usize>,
+    done_cache: FnvHashSet<O::Predicate>,
+    waiting_cache: FnvHashMap<O::Predicate, NodeIndex>,
+    cache_list: Vec<O::Predicate>,
+    snapshots: Vec<SnapshotData>,
+    scratch: Option<Vec<usize>>,
 }
 
 pub struct Snapshot {
     len: usize,
 }
 
-struct Tree<T> {
-    root: NodeIndex,
-    state: T,
-}
-
+#[derive(Debug)]
 struct Node<O> {
-    state: NodeState<O>,
+    obligation: O,
+    state: Cell<NodeState>,
+
+    // these both go *in the same direction*.
     parent: Option<NodeIndex>,
-    tree: TreeIndex,
+    dependants: Vec<NodeIndex>,
```

Comment:
(But I think I'd prefer to call them `children`, to be more analogous with `parent`)